test/jdk/java/foreign/TestCleaner.java

```diff
@@ -0,0 +1,135 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @modules java.base/jdk.internal.ref
+ *          jdk.incubator.foreign/jdk.incubator.foreign
+ * @run testng/othervm -Dforeign.restricted=permit TestCleaner
+ */
+
+import jdk.incubator.foreign.MemoryAccess;
+import jdk.incubator.foreign.MemoryAddress;
+import jdk.incubator.foreign.MemorySegment;
+import java.lang.ref.Cleaner;
+import jdk.internal.ref.CleanerFactory;
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import static org.testng.Assert.*;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class TestCleaner {
+
+    static class SegmentState {
+        private AtomicInteger cleanupCalls = new AtomicInteger(0);
+
+        void cleanup() {
+            cleanupCalls.incrementAndGet();
+        }
+
+        int cleanupCalls() {
+            return cleanupCalls.get();
+        }
+    }
+
+    @Test(dataProvider = "cleaners")
+    public void test(int n, Supplier<Cleaner> cleanerFactory, SegmentFunction segmentFunction) {
+        SegmentState segmentState = new SegmentState();
+        MemorySegment segment = makeSegment(segmentState);
+        // register cleaners before
+        for (int i = 0 ; i < n ; i++) {
+            segment.registerCleaner(cleanerFactory.get());
+        }
+        segment = segmentFunction.apply(segment);
+        if (segment.isAlive()) {
+            // also register cleaners after
+            for (int i = 0; i < n; i++) {
+                segment.registerCleaner(cleanerFactory.get());
+            }
+        }
+        //check that cleanup has not been called by any cleaner yet!
+        assertEquals(segmentState.cleanupCalls(), segment.isAlive() ? 0 : 1);
+        segment = null;
+        while (segmentState.cleanupCalls() == 0) {
+            byte[] b = new byte[100];
+            System.gc();
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException ex) {
+                throw new AssertionError(ex);
+            }
+        }
+        assertEquals(segmentState.cleanupCalls(), 1);
+    }
+
+    MemorySegment makeSegment(SegmentState segmentState) {
+        return MemorySegment.ofNativeRestricted(MemoryAddress.NULL, 10, Thread.currentThread(), segmentState::cleanup, null);
+    }
+
+    enum SegmentFunction implements Function<MemorySegment, MemorySegment> {
+        IDENTITY(Function.identity()),
+        CLOSE(s -> { s.close(); return s; }),
```

Comment:
Won't that just skip cleaner registration, making the test redundant for that case? I guess for this case it should be called after registration?

Reply:
Not really - note that the function is called after registration:
```
MemorySegment segment = makeSegment(segmentState);
        // register cleaners before
        for (int i = 0 ; i < n ; i++) {
            segment.registerCleaner(cleanerFactory.get());
        }
        segment = segmentFunction.apply(segment); // <-------------------------
```
So, the test here is useful: it tests that the cleaner registered before close() can't really call the cleanup action multiple times if the segment has also been closed explicitly.

Reply:
Oh, doh!, i missed the code above.